elide-core/src/main/java/com/yahoo/elide/security/PermissionManager.java

```diff
@@ -0,0 +1,500 @@
+                log.debug("Forbidden access at entity-level.");
+                throw e;
+            }
+        }
+
+        // If that succeeds, queue up our commit checks
+        if (comChecks.length > 0) {
+            commitChecks.add(() -> {
+                runPermissionChecks(comChecks, mode, resource, changeSpec);
+                return null;
+            });
+        }
+    }
+
+    /**
+     * Check for permissions on a class and its fields.
+     *
+     * @param resource resource
+     * @param changeSpec change spec
+     * @param annotationClass annotation class
+     * @param <A> type parameter
+     */
+    public <A extends Annotation> void checkFieldAwarePermissions(PersistentResource<?> resource,
+                                                                  ChangeSpec changeSpec,
+                                                                  Class<A> annotationClass) {
+        EntityDictionary dictionary = resource.getDictionary();
+
+        // First gather all checks and modes
+        Class<? extends Check>[] classOpChecks = null;
+        Class<? extends Check>[] classComChecks = null;
+        final ArrayList<Class<? extends Check>[]> fieldOpChecks = new ArrayList<>();
+        final ArrayList<Class<? extends Check>[]> fieldComChecks = new ArrayList<>();
+        CheckMode classCheckMode = null;
+        final ArrayList<CheckMode> fieldCheckModes = new ArrayList<>();
+
+        A annotation = dictionary.getAnnotation(resource, annotationClass);
+        if (annotation != null) {
+            ExtractedChecks extracted = extractChecks(annotationClass, annotation);
+            classOpChecks = extracted.getOperationChecks();
+            classComChecks = extracted.getCommitChecks();
+            classCheckMode = extracted.getCheckMode();
+        }
+
+        // Get all fields
+        Class<?> entityClass = resource.getResourceClass();
+        List<String> attributes = dictionary.getAttributes(entityClass);
+        List<String> relationships = dictionary.getRelationships(entityClass);
+        List<String> fields = (attributes != null) ? new ArrayList<>(dictionary.getAttributes(entityClass))
+                                                   : new ArrayList<>();
+        if (relationships != null) {
+            fields.addAll(relationships);
+        }
+
+        for (String field : fields) {
+            annotation = dictionary.getAttributeOrRelationAnnotation(entityClass, annotationClass, field);
+            if (annotation == null) {
+                continue;
+            }
+            ExtractedChecks extracted = extractChecks(annotationClass, annotation);
+            fieldOpChecks.add(extracted.getOperationChecks());
+            fieldComChecks.add(extracted.getCommitChecks());
+            fieldCheckModes.add(extracted.getCheckMode());
+        }
+
+        boolean hasDeferredChecks = (!isEmptyCheckArray(classComChecks) || !isEmptyListOfChecks(fieldComChecks))
```

Comment:
I still feel like we should be solely paying attention to type of check - not the type of permission.